### PR TITLE
check C/C++ struct member

### LINKED ISSUE #1065: Add check for structure member
This issue was originally created at: 2005-02-11 10:05:06.
This issue was reported by: `hsn`.
hsn said at 2005-02-11 10:05:06
>Add framework function for checking for specified member of some structure. I need to check struct dirent members.

issues@scons said at 2005-02-11 10:05:06
>Converted from SourceForge tracker item 1120953

gregnoel said at 2009-02-06 05:48:37
>Bug party triage.  Gnu configure calls it AC_CHECK_MEMBER; it ought to be a fairly simple wrapper around TryLink().

Edited: link updated:  https://www.gnu.org/software/autoconf/manual/autoconf-2.71/html_node/Generic-Structures.html

### PR DESCRIPTION
**Pull request by yuzhichang**
Added CheckMember. Close #1065.

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: SCons/Conftest.py [hunk 1]
```diff
@@ -568,6 +568,63 @@ def CheckDeclaration(context, symbol, includes = None, language = None):
...
+
+int main(void) {
+  if (sizeof ((%(aggregate)s *) 0)->%(member)s)
+    return 0;
+}''' % { 'include': includetext,
+         'header': header,
+         'aggregate': aggregate,
+         'member': member }
+
+    ret = context.BuildProg(text, suffix)
```
Comment (yuzhichang) [inline SCons/Conftest.py:617]: I don't see why context.BuildProg toggle the return value of TryBuild, _YesNoResult toggle it again, SConf.CheckHeader toggle it another time. It's confusing and error-prone. I would suggest unify them(except for CheckTypeSize, CheckProg) like this: return True for success, False for failure.

Comment (mwichmann) [reply SCons/Conftest.py:617]: I've been trying to document this (unfortunate) behavior, as this subsystem deals in both OS-level return values (0 success, nonzero fail) and Python return values (0 False, 1 True).  I don't think we can easily clean it up, but agree it's a real pain to follow.

Comment (yuzhichang) [reply SCons/Conftest.py:617]: I added the 2nd commit to unify the return value. My change passed the test:
```
$ python3 ./runtest.py test/Configure/
$ python3 ./runtest.py SCons/SConfTests.py
```

**Comment by bdbaddog** [comment]
Please add a blurb in CHANGES.txt and RELEASE.txt

If you can add docs for this checker that'd be great, we can help.

**Comment by yuzhichang** [comment]
Done according to your suggestion.

**Comment by mwichmann** [comment]
This has managed to break one existing test (not the added one) - fails on all four Windows builds:
```console
======================================================================
FAIL: test_CheckDeclaration (__main__.SConfTestCase)
Test SConf.CheckDeclaration()
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\scons\SCons\SConfTests.py", line 724, in test_CheckDeclaration
    assert not r, "unknown language was supported ??"
AssertionError: unknown language was supported ??
```

**Comment by yuzhichang** [comment]
@mwichmann I'm not able to run any test on windows 10.
```
eoitek@DESKTOP-JUE0AAG C:\Users\eoitek\scons>python ./runtest.py test/Configure/    
1/25 (4.00%) "C:\\Program Files\\Python310\\python.exe" test/Configure/Action-error.py
['"C:\\\\Program Files\\\\Python310\\\\python.exe"', 'test/Configure/Action-error.py']    --- This is the debug output I added to runtest.py:328
Traceback (most recent call last):
  File "C:\Users\eoitek\scons\runtest.py", line 853, in <module>
    run_test(t, io_lock=None, run_async=False)
  File "C:\Users\eoitek\scons\runtest.py", line 817, in run_test
    t.execute(env)
  File "C:\Users\eoitek\scons\runtest.py", line 425, in execute
    self.stderr, self.stdout, s = spawn_it(self.command_args, env)
  File "C:\Users\eoitek\scons\runtest.py", line 329, in spawn_it
    cp = subprocess.run(command_args, shell=False, env=env)
  File "C:\Program Files\Python310\lib\subprocess.py", line 501, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Program Files\Python310\lib\subprocess.py", line 966, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Program Files\Python310\lib\subprocess.py", line 1435, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
PermissionError: [WinError 5] 拒绝访问。

```

Login with administrator and run tests, no luck.

**Comment by bdbaddog** [comment]
@yuzhichang  - can you translate the Permission error message at the end?
Could this be caused by some anti-virus you have on your system?

**Comment by yuzhichang** [comment]
@bdbaddog I can build C/C++ projects with SCons on windows 10 and 2008, but not able to run scons tests on them. 
Firewalls are disabled, no anti-virus software are installed. Here's the error message.

```
eoitek@WIN10 C:\Users\eoitek\scons>python runtest.py test/Configure
1/25 (4.00%) "C:\\Program Files\\Python310\\python.exe" test/Configure\Action-error.py
Traceback (most recent call last):
  File "C:\Users\eoitek\scons\runtest.py", line 852, in <module>       
    run_test(t, io_lock=None, run_async=False)
  File "C:\Users\eoitek\scons\runtest.py", line 816, in run_test       
    t.execute(env)
  File "C:\Users\eoitek\scons\runtest.py", line 424, in execute        
    self.stderr, self.stdout, s = spawn_it(self.command_args, env)     
  File "C:\Users\eoitek\scons\runtest.py", line 328, in spawn_it       
    cp = subprocess.run(command_args, shell=False, env=env)
  File "C:\Program Files\Python310\lib\subprocess.py", line 501, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Program Files\Python310\lib\subprocess.py", line 966, in __init__       
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Program Files\Python310\lib\subprocess.py", line 1435, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
PermissionError: [WinError 5] Access is denied

eoitek@WIN2008SRV C:\Users\eoitek\scons>python runtest.py test/Configure                                                        
1/25 (4.00%) "C:\\Program Files\\Python37\\python.exe" test/Configure\Action-error.py                                           
Traceback (most recent call last):                                                                                              
  File "runtest.py", line 852, in <module>                                                                                      
    run_test(t, io_lock=None, run_async=False)                                                                                  
  File "runtest.py", line 816, in run_test                                                                                      
    t.execute(env)                                                                                                              
  File "runtest.py", line 424, in execute                                                                                       
    self.stderr, self.stdout, s = spawn_it(self.command_args, env)                                                              
  File "runtest.py", line 328, in spawn_it                                                                                      
    cp = subprocess.run(command_args, shell=False, env=env)                                                                     
  File "C:\Program Files\Python37\lib\subprocess.py", line 488, in run                                                          
    with Popen(*popenargs, **kwargs) as process:                                                                                
  File "C:\Program Files\Python37\lib\subprocess.py", line 800, in __init__                                                     
    restore_signals, start_new_session)                                                                                         
  File "C:\Program Files\Python37\lib\subprocess.py", line 1207, in _execute_child                                              
    startupinfo)                                                                                                                
PermissionError: [WinError 5] Access is denied                                                                                  
```

Can you help to fix test_CheckDeclaration failure since my env is unavailable?

**Comment by bdbaddog** [comment]
That's very odd can you try:

`python runtest.py --verbose 2 test/Configure\Action-error.py`

**Comment by yuzhichang** [comment]
```
eoitek@WIN10 C:\Users\eoitek\scons>python runtest.py --verbose 2 test\Configure\Action-error.py 
1/1 (100.00%) "C:\\Program Files\\Python310\\python.exe" test\Configure\Action-error.py
Traceback (most recent call last):
  File "C:\Users\eoitek\scons\runtest.py", line 852, in <module>
    run_test(t, io_lock=None, run_async=False)
  File "C:\Users\eoitek\scons\runtest.py", line 816, in run_test
    t.execute(env)
  File "C:\Users\eoitek\scons\runtest.py", line 424, in execute
    self.stderr, self.stdout, s = spawn_it(self.command_args, env)
  File "C:\Users\eoitek\scons\runtest.py", line 328, in spawn_it
    cp = subprocess.run(command_args, shell=False, env=env)
  File "C:\Program Files\Python310\lib\subprocess.py", line 501, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Program Files\Python310\lib\subprocess.py", line 966, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Program Files\Python310\lib\subprocess.py", line 1435, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
PermissionError: [WinError 5] Access is denied

```

**Comment by mwichmann** [comment]
This is definitely odd, have no clue, and StackOverflow isn't being directly helpful either, though this error has been seen. 

```python
  File "C:\Program Files\Python310\lib\subprocess.py", line 1435, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
PermissionError: [WinError 5] Access is denied
```

That's not us (directly), that's subprocess unable to start a process.

You said it's all tests, not just this one, right?

**Comment by yuzhichang** [comment]
Right. So I'm unable to fix the broken `test_CheckDeclaration` you said. Can you fix it instead?

**Comment by bdbaddog** [comment]
Weird.  I pushed a sider fix. But I cannot repro the SConfTests.py failure with either 3.8 or 3.10 on win 10.

```(venv310) PS C:\Users\Bill\devel\scons\git\yuzhichang> python .\runtest.py .\SCons\SConfTests.py
1/1 (100.00%) C:\\Users\\Bill\\devel\\scons\\venv310\\Scripts\\python.exe SCons\SConfTests.py
<frozen importlib._bootstrap>:283: DeprecationWarning: the load_module() method is deprecated and slated for removal in Python 3.12; use exec_module() instead
......................
----------------------------------------------------------------------
Ran 22 tests in 58.700s

OK
(venv310) PS C:\Users\Bill\devel\scons\git\yuzhichang> deactivate
PS C:\Users\Bill\devel\scons\git\yuzhichang> ..\..\venv\Scripts\python.exe .\runtest.py .\SCons\SConfTests.py
1/1 (100.00%) C:\\Users\\Bill\\devel\\scons\\venv\\Scripts\\python.exe SCons\SConfTests.py
......................
----------------------------------------------------------------------
Ran 22 tests in 57.934s

OK```

**Comment by yuzhichang** [comment]
'Linux Testing' has one failure: `test/Dir/source.py`.
This test is flaky. It passes on CentOS, failes on Ubuntu based OS.

| scons branch | os | pass? |
|-|-|-|
| check_member_struct | CentOS 7.9 | Pass |
| check_member_struct | Mint 20.3 | Fail |
| master | CentOS 7.9 | Pass |
| master | Mint 20.3 | Fail |

The failure text is as following:
```
zhichyu@leno:~/scons$ python3 runtest.py test/Dir/source.py
1/1 (100.00%) /usr/bin/python3 test/Dir/source.py
STDOUT =========================================================================
1,6c1,5
< scons:\ Reading\ SConscript\ files\ \.\.\.\
< scons:\ done\ reading\ SConscript\ files\.\
< scons:\ Building\ targets\ \.\.\.\
< ((?!scons: `tstamp\.out' is up to date.)[^
< ]*
< )*scons:\ done\ building\ targets\.\
---
> scons: Reading SConscript files ...
> scons: done reading SConscript files.
> scons: Building targets ...
> scons: `tstamp.out' is up to date.
> scons: done building targets.
FAILED test of /home/zhichyu/scons/scripts/scons.py
	at line 659 of /home/zhichyu/scons/testing/framework/TestCommon.py (_complete)
	from line 754 of /home/zhichyu/scons/testing/framework/TestCommon.py (run)
	from line 449 of /home/zhichyu/scons/testing/framework/TestSCons.py (run)
	from line 505 of /home/zhichyu/scons/testing/framework/TestSCons.py (not_up_to_date)
	from line 110 of test/Dir/source.py

```

**Comment by bdbaddog** [comment]
Not sure we can take the change in return values to the configure context functions. Some of them are documented to return a string contain an error message on failure..

See: [configure contexts from manpage](https://scons.org/doc/production/HTML/scons-man.html#configure_contexts)

To change that we'd have to go through a deprecation cycle.

@mwichmann  - thoughts?

**Comment by mwichmann** [comment]
I'm not too worried about changing the returns a bit - I think the documentation on that happening (could return a string on fail) came from me prospecting in the code fairly recently, I'd want to check an earlier doc version to see if we ever said that before.  But it would be a lot easier to review and accept the core change here - the new member check - if there wasn't so much other change.  So my recommendation would be to split into separate PRs, I know when I arrived to take a glance it drew a "okay, I'm going to put this off for a while".  I'll defer that to @bdbaddog as the maintainer, that's who would have to accept.

**Comment by bdbaddog** [comment]
Yes. Ideally there'd be a PR for the new checker, and a separate one to clean up the return values..

**Comment by bdbaddog** [comment]
merging just the checker is a no brainer. Changing the return value requires more thorough review (as detailed above)

**Comment by bdbaddog** [comment]
Closing.
The new checker is added in #4122 

The changing the return value changes can be readdressed at a later point, but I'm not sure we can universally make such change as some checkers need to be able to return more than true/false.

**Pull request closed without merging**